pex/pip/installation.py

```diff
@@ -77,15 +79,63 @@ def _pip_installation(
     return Pip(pip=pip_venv, version=version, pip_cache=pip_cache)
 
 
-def _vendored_installation(interpreter=None):
-    # type: (Optional[PythonInterpreter]) -> Pip
+def _fingerprint(requirements):
+    # type: (Tuple[Requirement, ...]) -> str
+    if not requirements:
+        return "no-extra-requirements"
+    return hashlib.sha1("\n".join(sorted(map(str, requirements))).encode("utf-8")).hexdigest()
+
+
+def _vendored_installation(
+    interpreter=None,  # type: Optional[PythonInterpreter]
+    resolver=None,  # type: Optional[Resolver]
+    extra_requirements=(),  # type: Tuple[Requirement, ...]
+):
+    # type: (...) -> Pip
+
+    def expose_vendored():
+        # type: () -> Iterator[str]
+        return third_party.expose(("pip", "setuptools"), interpreter=interpreter)
+
+    if not extra_requirements:
+        return _pip_installation(
+            version=PipVersion.VENDORED,
+            iter_distribution_locations=expose_vendored,
+            interpreter=interpreter,
+            fingerprint=_fingerprint(extra_requirements),
+        )
+
+    if not resolver:
+        raise ValueError(
+            "A resolver is required to install extra {requirements} for vendored Pip: "
+            "{extra_requirements}".format(
+                requirements=pluralize(extra_requirements, "requirement"),
+                extra_requirements=" ".join(map(str, extra_requirements)),
+            )
+        )
+
+    # This indirection works around MyPy type inference failing to see that
+    # `iter_distribution_locations` is only successfully defined when resolve is not None.
+    extra_requirement_resolver = resolver
+
+    def iter_distribution_locations():
+        # type: () -> Iterator[str]
+        for location in expose_vendored():
+            yield location
+
+        for resolved_distribution in extra_requirement_resolver.resolve_requirements(
+            requirements=tuple(map(str, extra_requirements)),
+            targets=Targets.from_target(LocalInterpreter.create(interpreter)),
+            pip_version=PipVersion.VENDORED,
+            extra_resolver_requirements=(),
+        ).distributions:
+            yield resolved_distribution.distribution.location
 
     return _pip_installation(
         version=PipVersion.VENDORED,
-        iter_distribution_locations=lambda: third_party.expose(
-            ("pip", "setuptools"), interpreter=interpreter
-        ),
+        iter_distribution_locations=iter_distribution_locations,
         interpreter=interpreter,
+        fingerprint=_fingerprint(extra_requirements),
     )
```

Comment:
N.B.: The PR was motivated by Pip keyring but is more generic, allowing extra pip requirements for any reason under the sun that comes up. As such, even though keyring support does not work with vendored Pip, `--extra-pip-requirement` should work with vendored Pip.